File: src/main/java/org/example/kalkulationsprogramm/controller/LohnStammdatenController.java
package org.example.kalkulationsprogramm.controller;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.example.kalkulationsprogramm.dto.GewerkDto;
import org.example.kalkulationsprogramm.dto.KrankenkasseDto;
import org.example.kalkulationsprogramm.dto.SvSatzDto;
import org.example.kalkulationsprogramm.service.GewerkService;
import org.example.kalkulationsprogramm.service.KrankenkasseService;
import org.example.kalkulationsprogramm.service.SvSatzService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

/**
 * Stammdaten fuer Lohn-/Sozialversicherungs-Berechnung:
 *   - Krankenkassen (mit Zusatzbeitrag)
 *   - SV-Saetze (KV/PV/RV/AV/Minijob/Umlagen)
 *   - Gewerke + zugeordnete Berufsgenossenschaft (Unfallversicherung)
 *
 * Alle drei sind im Frontend editierbar - die Migrationen liefern nur den
 * Initial-Seed.
 */
@RestController
@RequestMapping("/api/lohn-stammdaten")
@RequiredArgsConstructor
public class LohnStammdatenController {

    private final KrankenkasseService krankenkasseService;
    private final SvSatzService svSatzService;
    private final GewerkService gewerkService;

    // ==================== KRANKENKASSEN ====================

    @GetMapping("/krankenkassen")
    public ResponseEntity<List<KrankenkasseDto>> getKrankenkassen(
            @RequestParam(name = "nurAktive", required = false, defaultValue = "false") boolean nurAktive) {
        return ResponseEntity.ok(nurAktive ? krankenkasseService.findAktiv() : krankenkasseService.findAll());
    }

    @PostMapping("/krankenkassen")
    public ResponseEntity<?> erstelleKrankenkasse(@Valid @RequestBody KrankenkasseDto dto) {
        try {
            dto.setId(null);
            return ResponseEntity.ok(krankenkasseService.save(dto));
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.badRequest().body(Map.of("message", ex.getMessage()));
        }
    }

    @PutMapping("/krankenkassen/{id}")
    public ResponseEntity<?> aktualisiereKrankenkasse(@PathVariable Long id, @Valid @RequestBody KrankenkasseDto dto) {
        try {
            dto.setId(id);
            return ResponseEntity.ok(krankenkasseService.save(dto));
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.badRequest().body(Map.of("message", ex.getMessage()));
        }
    }

    @DeleteMapping("/krankenkassen/{id}")
    public ResponseEntity<Void> loescheKrankenkasse(@PathVariable Long id) {
        krankenkasseService.delete(id);
        return ResponseEntity.noContent().build();
    }

    // ==================== SV-SAETZE ====================

    @GetMapping("/sv-saetze")
    public ResponseEntity<List<SvSatzDto>> getSvSaetze() {
        return ResponseEntity.ok(svSatzService.findAll());
    }

    @PostMapping("/sv-saetze")
    public ResponseEntity<?> erstelleSvSatz(@Valid @RequestBody SvSatzDto dto) {
        try {
            dto.setId(null);
            return ResponseEntity.ok(svSatzService.save(dto));
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.badRequest().body(Map.of("message", ex.getMessage()));
        }
    }

    @PutMapping("/sv-saetze/{id}")
    public ResponseEntity<?> aktualisiereSvSatz(@PathVariable Long id, @Valid @RequestBody SvSatzDto dto) {
        try {
            dto.setId(id);
            return ResponseEntity.ok(svSatzService.save(dto));
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.badRequest().body(Map.of("message", ex.getMessage()));
        }
    }

    @DeleteMapping("/sv-saetze/{id}")
    public ResponseEntity<Void> loescheSvSatz(@PathVariable Long id) {
        svSatzService.delete(id);
        return ResponseEntity.noContent().build();
    }

    // ==================== GEWERKE ====================

    @GetMapping("/gewerke")
    public ResponseEntity<List<GewerkDto>> getGewerke(
            @RequestParam(name = "nurAktive", required = false, defaultValue = "false") boolean nurAktive) {
        return ResponseEntity.ok(nurAktive ? gewerkService.findAktiv() : gewerkService.findAll());
    }

    @PostMapping("/gewerke")
    public ResponseEntity<?> erstelleGewerk(@Valid @RequestBody GewerkDto dto) {
        try {
            dto.setId(null);
            return ResponseEntity.ok(gewerkService.save(dto));
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.badRequest().body(Map.of("message", ex.getMessage()));
        }
    }

    @PutMapping("/gewerke/{id}")
    public ResponseEntity<?> aktualisiereGewerk(@PathVariable Long id, @Valid @RequestBody GewerkDto dto) {
        try {
            dto.setId(id);
            return ResponseEntity.ok(gewerkService.save(dto));
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.badRequest().body(Map.of("message", ex.getMessage()));
        }
    }

    @DeleteMapping("/gewerke/{id}")
    public ResponseEntity<Void> loescheGewerk(@PathVariable Long id) {
        gewerkService.delete(id);
        return ResponseEntity.noContent().build();
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/controller/LohnStammdatenController.kt
package org.example.kalkulationsprogramm.controller

import jakarta.validation.Valid
import org.example.kalkulationsprogramm.dto.GewerkDto
import org.example.kalkulationsprogramm.dto.KrankenkasseDto
import org.example.kalkulationsprogramm.dto.SvSatzDto
import org.example.kalkulationsprogramm.service.GewerkService
import org.example.kalkulationsprogramm.service.KrankenkasseService
import org.example.kalkulationsprogramm.service.SvSatzService
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/lohn-stammdaten")
class LohnStammdatenController(
    private val krankenkasseService: KrankenkasseService,
    private val svSatzService: SvSatzService,
    private val gewerkService: GewerkService,
) {
    @GetMapping("/krankenkassen")
    fun getKrankenkassen(
        @RequestParam(name = "nurAktive", required = false, defaultValue = "false") nurAktive: Boolean,
    ): ResponseEntity<List<KrankenkasseDto>> {
        return ResponseEntity.ok(if (nurAktive) krankenkasseService.findAktiv() else krankenkasseService.findAll())
    }

    @PostMapping("/krankenkassen")
    fun erstelleKrankenkasse(@Valid @RequestBody dto: KrankenkasseDto): ResponseEntity<Any> {
        return try {
            dto.id = null
            ResponseEntity.ok(krankenkasseService.save(dto))
        } catch (ex: IllegalArgumentException) {
            ResponseEntity.badRequest().body(mapOf("message" to ex.message))
        }
    }

    @PutMapping("/krankenkassen/{id}")
    fun aktualisiereKrankenkasse(@PathVariable id: Long, @Valid @RequestBody dto: KrankenkasseDto): ResponseEntity<Any> {
        return try {
            dto.id = id
            ResponseEntity.ok(krankenkasseService.save(dto))
        } catch (ex: IllegalArgumentException) {
            ResponseEntity.badRequest().body(mapOf("message" to ex.message))
        }
    }

    @DeleteMapping("/krankenkassen/{id}")
    fun loescheKrankenkasse(@PathVariable id: Long): ResponseEntity<Void> {
        krankenkasseService.delete(id)
        return ResponseEntity.noContent().build()
    }

    @GetMapping("/sv-saetze")
    fun getSvSaetze(): ResponseEntity<List<SvSatzDto>> = ResponseEntity.ok(svSatzService.findAll())

    @PostMapping("/sv-saetze")
    fun erstelleSvSatz(@Valid @RequestBody dto: SvSatzDto): ResponseEntity<Any> {
        return try {
            dto.id = null
            ResponseEntity.ok(svSatzService.save(dto))
        } catch (ex: IllegalArgumentException) {
            ResponseEntity.badRequest().body(mapOf("message" to ex.message))
        }
    }

    @PutMapping("/sv-saetze/{id}")
    fun aktualisiereSvSatz(@PathVariable id: Long, @Valid @RequestBody dto: SvSatzDto): ResponseEntity<Any> {
        return try {
            dto.id = id
            ResponseEntity.ok(svSatzService.save(dto))
        } catch (ex: IllegalArgumentException) {
            ResponseEntity.badRequest().body(mapOf("message" to ex.message))
        }
    }

    @DeleteMapping("/sv-saetze/{id}")
    fun loescheSvSatz(@PathVariable id: Long): ResponseEntity<Void> {
        svSatzService.delete(id)
        return ResponseEntity.noContent().build()
    }

    @GetMapping("/gewerke")
    fun getGewerke(
        @RequestParam(name = "nurAktive", required = false, defaultValue = "false") nurAktive: Boolean,
    ): ResponseEntity<List<GewerkDto>> {
        return ResponseEntity.ok(if (nurAktive) gewerkService.findAktiv() else gewerkService.findAll())
    }

    @PostMapping("/gewerke")
    fun erstelleGewerk(@Valid @RequestBody dto: GewerkDto): ResponseEntity<Any> {
        return try {
            dto.id = null
            ResponseEntity.ok(gewerkService.save(dto))
        } catch (ex: IllegalArgumentException) {
            ResponseEntity.badRequest().body(mapOf("message" to ex.message))
        }
    }

    @PutMapping("/gewerke/{id}")
    fun aktualisiereGewerk(@PathVariable id: Long, @Valid @RequestBody dto: GewerkDto): ResponseEntity<Any> {
        return try {
            dto.id = id
            ResponseEntity.ok(gewerkService.save(dto))
        } catch (ex: IllegalArgumentException) {
            ResponseEntity.badRequest().body(mapOf("message" to ex.message))
        }
    }

    @DeleteMapping("/gewerke/{id}")
    fun loescheGewerk(@PathVariable id: Long): ResponseEntity<Void> {
        gewerkService.delete(id)
        return ResponseEntity.noContent().build()
    }
}
